cpp/csp/adapters/websocket/WebsocketEndpoint.h

```diff
@@ -1,84 +1,367 @@
 #ifndef _IN_CSP_ADAPTERS_WEBSOCKETS_ENDPOINT_H
 #define _IN_CSP_ADAPTERS_WEBSOCKETS_ENDPOINT_H
 
-// need a base -> TLS, base -> No TLS
-#include <csp/adapters/websocket/ClientInputAdapter.h>
-#include <websocketpp/config/asio_client.hpp>
-#include <websocketpp/config/asio_no_tls_client.hpp>
-#include <websocketpp/client.hpp>
+#include <boost/beast/core.hpp>
+#include <boost/beast/ssl.hpp>
+#include <boost/beast/websocket.hpp>
+#include <boost/asio/strand.hpp>
+
+#include <csp/engine/Dictionary.h>
+#include <csp/core/Exception.h>
+
+#include <cstdlib>
 #include <functional>
+#include <iostream>
+#include <string>
 
 namespace csp::adapters::websocket {
+using namespace csp;
 
-using on_message_cb = std::function<void(std::string)>;
-using on_send_fail_cb = std::function<void(const std::string&)>;
-using void_cb = std::function<void()>;
-
-using message_ptr = websocketpp::config::core_client::message_type::ptr;
+namespace beast = boost::beast;         // from <boost/beast.hpp>
+namespace http = beast::http;           // from <boost/beast/http.hpp>
+namespace net = boost::asio;            // from <boost/asio.hpp>
+namespace ssl = boost::asio::ssl;       // from <boost/asio/ssl.hpp>
+namespace websocket = beast::websocket; // from <boost/beast/websocket.hpp>
+using tcp = boost::asio::ip::tcp;       // from <boost/asio/ip/tcp.hpp>
 
+using string_cb = std::function<void(const std::string&)>;
+using void_cb = std::function<void()>;
 
-class WebsocketEndpointBase {
+class BaseWebsocketSession {
 public:
-    WebsocketEndpointBase(csp::Dictionary properties);
-    virtual ~WebsocketEndpointBase() { };
+    BaseWebsocketSession(
+        net::io_context& ioc,
+        Dictionary* properties,
+        void_cb& on_open,
+        string_cb& on_fail,
+        string_cb& on_message,
+        void_cb& on_close,
+        string_cb& on_send_fail
+    ) : m_resolver( net::make_strand( ioc ) ),
+        m_properties( properties ), 
+        m_on_open( on_open ),
+        m_on_fail( on_fail ),
+        m_on_message( on_message ),
+        m_on_close( on_close ),
+        m_on_send_fail( on_send_fail )
+    { };
 
     virtual void run() { };
-    virtual void send(const std::string&){ };
-    virtual void close() { };
+    virtual void stop() { };
+    virtual void send( const std::string& ) { };
+    virtual void do_read() { };
+    void handle_message( beast::error_code& ec, std::size_t& bytes_transfered ) {
+        if( ec ) {
+            m_on_close();
+            return;
+        }
+        m_on_message(beast::buffers_to_string(m_buffer.data()));
```

Comment:
buffers_to_string looks pretty slow and unnecessary.
Should be able to pass the raw data through without copying/processing
https://boost.org/doc/libs/1_84_0/boost/beast/core/buffers_to_string.hpp

Reply:
Updated it to: static_cast<char const*>(beast::buffers_front(m_buffer.data()).data())